plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/PermitBackedRetryableFutureUtils.java

```diff
@@ -0,0 +1,328 @@
+        CompletableFuture<CompletableFuture<FutureType>> completableCompletable
+    ) {
+        return completableCompletable.thenCompose(Function.identity());
+    }
+
+    private CompletableFuture<FutureType> retry(
+        Throwable ex,
+        Function<Semaphore, CompletableFuture<FutureType>> futureSupplier,
+        Iterator<TimeValue> retryBackoffDelayIterator
+    ) {
+        if (!(ex instanceof RetryableException)) {
+            return CompletableFuture.failedFuture(ex);
+        }
+
+        RetryableException retryableException = (RetryableException) ex;
+        if (!retryBackoffDelayIterator.hasNext()) {
+            return CompletableFuture.failedFuture(ex);
+        }
+
+        return unwrap(
+            unwrap(
+                CompletableFuture.supplyAsync(
+                    () -> futureSupplier.apply(retryableException.semaphore),
+                    new DelayedExecutor(
+                        retryableException.retryBackoffDelayIterator.next().millis(),
+                        TimeUnit.MILLISECONDS,
+                        remoteTransferRetryPool,
+                        scheduler
+                    )
+                )
+            ).thenApply(CompletableFuture::completedFuture).exceptionally(t -> {
+                if (t instanceof RetryableException) {
+                    ex.addSuppressed(t);
+                    return retry(ex, futureSupplier, retryBackoffDelayIterator);
+                } else {
+                    ex.addSuppressed(t);
+                    return CompletableFuture.failedFuture(ex);
+                }
+            })
+        );
+    }
+
+    // Package access for testing
+    Semaphore acquirePermit(WritePriority writePriority, boolean isLowPriorityPermitsConsumable) throws InterruptedException {
+        // Try acquiring low priority permit or high priority permit immediately if available.
+        // Otherwise, we wait for low priority permit.
+        if (Objects.requireNonNull(writePriority) == WritePriority.LOW) {
+            if (lowPrioritySemaphore.tryAcquire()) {
+                return lowPrioritySemaphore;
+            } else if (highPrioritySemaphore.availablePermits() > 0.4 * highPriorityPermits && highPrioritySemaphore.tryAcquire()) {
+                return highPrioritySemaphore;
+            } else if (lowPrioritySemaphore.tryAcquire(5, TimeUnit.MINUTES)) {
+                return lowPrioritySemaphore;
+            }
+            return null;
+        }
+
+        // Try acquiring high priority permit or low priority permit immediately if available.
+        // Otherwise, we wait for high priority permit.
+        if (highPrioritySemaphore.tryAcquire()) {
+            return highPrioritySemaphore;
+        } else if (isLowPriorityPermitsConsumable && lowPrioritySemaphore.tryAcquire()) {
+            return lowPrioritySemaphore;
+        } else if (highPrioritySemaphore.tryAcquire(5, TimeUnit.MINUTES)) {
```

Comment:
The code around your comment tries to optimize utilization in case of high priority uploads by utilizing low priority permits if no low priority transfer is in progress. If it is in progress then break out from that if and we wait for high priority permit in the worst case. Waiting for a low priority permit also here would interfere with low priority uploads.